src/commands/watch.rs

```diff
@@ -145,116 +173,126 @@ impl WatchCommand {
-                .collect::<Vec<String>>()
-                .as_mut(),
-        );
-        tracing::info!(
-            "proceeding with command: {} {}",
-            spin_cmd,
-            spin_args.join(" ")
-        );
-        (spin_cmd, spin_args)
+        spin_args.append(&mut vec![String::from("-f"), manifest_path]);
+        if matches!(state, State::Running) {
+            spin_args.extend(up_args);
+        }
+        spin_args
     }
 
-    async fn generate_path_patterns(&self) -> Result<Vec<String>> {
-        // TODO: Spend some time cleaning up this function
+    async fn generate_filter_config(&self) -> Result<crate::watch_filter::Config> {
+        let app_dir = parent_dir(&self.app)?;
         let app_manifest = spin_loader::local::raw_manifest_from_file(&self.app)
             .await?
             .into_v1();
 
-        let watch_patterns = app_manifest
+        // We always want to watch the application manifest
+        let manifest_pattern = WatchPattern::new(
+            self.app
+                .to_str()
+                .with_context(|| format!("non-unicode manifest path {:?}", self.app))?
+                .to_owned(),
+            true,
+            app_dir.as_path(),
+        )?;
+
+        // We want to watch the source code if we aren't skipping the build step
+        let source_patterns = app_manifest
             .components
             .iter()
             .filter_map(|c| {
-                if let Some(build) = c.build.as_ref() {
-                    match build.watch.clone() {
-                        Some(watch) => {
-                            let Some(workdir) = build.workdir.clone() else {
-                                return Some(watch);
-                            };
-                            Some(
-                                watch
-                                    .iter()
-                                    .filter_map(|ws| workdir.join(ws).to_str().map(String::from))
-                                    .collect::<Vec<String>>(),
-                            )
-                        }
-                        None => {
-                            eprintln!(
-                                "You haven't configured what to watch for the component: '{}'. Learn how to configure Spin watch at https://developer.fermyon.com/common/cli-reference#watch",
-                                c.id
-                            );
-                            None
-                        }
-                    }
-                } else {
-                    // No build config for this component so lets watch the source instead
-                    if let RawModuleSource::FileReference(path) = &c.source {
-                        if let Some(string_path) = path.to_str() {
-                            return Some(vec![String::from(string_path)]);
-                        }
-                    }
-                    None
-                }
+                let build = c.build.as_ref()?;
```

Comment:
I'd strongly encourage you to move this lambda out to a named function ~so I can understand it~ to more visibly express intent.

Reply:
Done